src/System.Security.Permissions/tests/PrincipalPermissionTests.cs

```diff
@@ -0,0 +1,329 @@
+            SecurityElement se2 = new SecurityElement("IInvalidPermission", se.Text);
+            se2.AddAttribute("class", se.Attribute("class"));
+            se2.AddAttribute("version", se.Attribute("version"));
+            Assert.Throws<ArgumentException>(() => p.FromXml(se2));
+        }
+
+        [Fact]
+        public void FromXmlWrongVersion()
+        {
+            PrincipalPermission p = new PrincipalPermission(PermissionState.None);
+            SecurityElement se = p.ToXml();
+            // can't modify - so we create our own
+            SecurityElement se2 = new SecurityElement(se.Tag, se.Text);
+            se2.AddAttribute("class", se.Attribute("class"));
+            se2.AddAttribute("version", "2");
+            Assert.Throws<ArgumentException>(() => p.FromXml(se2));
+        }
+
+        [Fact]
+        public void FromXml()
+        {
+            PrincipalPermission p = new PrincipalPermission(PermissionState.None);
+            SecurityElement se = p.ToXml();
+            Assert.NotNull(se);
+
+            PrincipalPermission p2 = (PrincipalPermission)p.Copy();
+            p2.FromXml(se);
+            Assert.Equal(p.ToString(), p2.ToString());
+
+            string className = (string)se.Attributes["class"];
+            string version = (string)se.Attributes["version"];
+
+            SecurityElement se2 = new SecurityElement(se.Tag);
+            se2.AddAttribute("class", className);
+            se2.AddAttribute("version", version);
+            p2.FromXml(se2);
+
+            SecurityElement sec = new SecurityElement("Identity");
+            sec.AddAttribute("Authenticated", "true");
+            se2.AddChild(sec);
+            p2.FromXml(se2);
+            Assert.True(p2.IsUnrestricted());
+        }
+
+        [Fact]
+        public void UnionWithNull()
+        {
+            PrincipalPermission p1 = new PrincipalPermission("user", null);
+            PrincipalPermission p2 = null;
+            PrincipalPermission p3 = (PrincipalPermission)p1.Union(p2);
+            Assert.Equal(p1.ToXml().ToString(), p3.ToXml().ToString());
+        }
+
+        [Fact]
+        public void UnionWithUnrestricted()
+        {
+            PrincipalPermission p1 = new PrincipalPermission(PermissionState.Unrestricted);
+            PrincipalPermission p2 = new PrincipalPermission("user", "role");
+            PrincipalPermission p3 = (PrincipalPermission)p1.Union(p2);
+            Assert.True(p3.IsUnrestricted());
+            p3 = (PrincipalPermission)p2.Union(p1);
+            Assert.True(p3.IsUnrestricted());
+        }
+
+        [Fact]
+        public void Union()
+        {
+            PrincipalPermission p1 = new PrincipalPermission("user A", "role A");
+            PrincipalPermission p2 = new PrincipalPermission("user B", "role B", false);
+            PrincipalPermission p3 = (PrincipalPermission)p1.Union(p2);
+            Assert.True(p3.ToString().IndexOf("user A") >= 0);
```

Comment:
string unionString = p3.ToString();
Assert.Contains("user A", unionString);

Also, don't forget the roles.